Adding several unit tests to improve coverage.

mgjarrett opened the pull request:
I selected files currently in the top 20 of missed lines on coveralls and skimmed through them to determine which had low-hanging fruit for improving test coverage statistics. I added unit tests to the following files:

armi/utils/tests/test_utils.py
armi/reactor/tests/test_reactors.py
armi/reactor/tests/test_blocks.py
armi/reactor/tests/test_zones.py

I tended towards smaller bites: unit tests for small functions with only a couple of relevant lines that could be easily implemented giving the existing test infrastructure and example objects that were already set up for other tests. I considered writing tests for more complicated functions that would have likely required setting up new data (e.g., burnup/power/flux summary functions that require spatial distributions) but generally stayed away from those.

I fixed three small bugs that I discovered while writing the tests:

1) armi/reactor/blocks.py: In the class ThRZBlock, the thetaOuter() method returned the min() of the "outer_theta" dimensions, but it seems like it should return the max() based on the name and comments.
2) armi/reactor/zones.py: The Zones method _buildAssemTypeZones() is supposed to create and name zones based on the assembly type (i.e., "igniter fuel", "lta fuel", etc.) but it was instead using the assembly name attribute, derived from the location ("A0010", "A0016", etc.). As a result, the assemblies were not correctly grouped into zones based on type.
3) armi/reactor/converters/geometryConverters.py: In the class BlockNumberModifier, the convert() method calls SetNumberOfBlocks, and passed it a string blockType (e.g., "clad") where it expects a flag (e.g., "Flags.CLAD"). Fixed with a call to Flags.fromString().

P.S. This is my first pull request. I used black and tried to keep naming and style in line with the surrounding code, but I wouldn't be surprised if I missed some subtleties (or even some not-so-subtleties). The commits are pretty atomic, so I'm hoping it will be easy to go back and revise them as necessary.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/terrapower/armi?pullRequest=253) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/terrapower/armi?pullRequest=253) before we can accept your contribution.<br/><hr/>**Michael Jarrett** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/terrapower/armi?pullRequest=253) it.</sub>

ntouran commented on armi/utils/tests/test_utils.py:
```diff
@@ -36,6 +36,9 @@ def test_parabola(self):
         self.assertEqual(b, 0.0)
         self.assertEqual(c, 1.0)
 
+        with self.assertRaises(Exception):
```
probably should get more specific than `Exception`. I get a `LinAlgError: Singular matrix`, so would `assertRaises(numpy.linalg.LinAlgError)` be better? Meh, maybe not. We're really just trying to make sure it fails and probably shouldn't be too sensitive to the internals of numpy. Ok you can leave it.

mgjarrett replied:
I didn't want to modify the actual code, which just has a "raise" statement with no specific exception/error type given.

ntouran commented on armi/utils/tests/test_utils.py:
```diff
@@ -202,6 +202,14 @@ def test_slantSplit(self):
         self.assertListEqual(x1, [1.0, 2.0, 3.0, 4.0])
         self.assertListEqual(x2, [4.0, 3.0, 2.0, 1.0])
 
+    def test_newtonsMethod(self):
```
Good call on leaving the cs out as we are hoping to get rid of most things that need some kind of global CS in our roadmap.

ntouran commented on armi/utils/tests/test_utils.py:
```diff
@@ -218,6 +220,12 @@ def test_minimizeScalarFunc(self):
+        a = ["Foo", "Bar"]
+        b = [1, 2, 3]
+        utils.prependToList(a, b)
+        self.assertTrue(a == [1, 2, 3, "Foo", "Bar"])
```
Does `self.assertEqual(a, [1, 2, 3, "Foo", "Bar"])` work here?

ntouran commented on armi/utils/tests/test_utils.py:
```diff
@@ -232,22 +232,21 @@ def test_prependToList(self):
         b = [1, 2, 3]
         utils.prependToList(a, b)
-        self.assertTrue(a == [1, 2, 3, "Foo", "Bar"])
+        self.assertListEqual(a, [1, 2, 3, "Foo", "Bar"])
```
Ah that answers my previous question.

mgjarrett replied:
I tried to clean up the commit history after the fact but it looks like I missed this.

ntouran commented on armi/utils/tests/test_utils.py:
```diff
@@ -17,6 +17,7 @@
 import unittest
 import math
+import os
 import numpy as np
```
should leave the single blank line between python built in imports and third party ones.

ntouran commented on armi/utils/tests/test_utils.py:
```diff
@@ -250,6 +251,20 @@ def test_convertToSlice(self):
             slice1 = utils.convertToSlice("slice")
 
+    def test_plotMatrix(self):
+        fname = "F"
```
I recommend using a more unique filename here. We run into trouble in parallel with os.remove locks sometimes in the off chance some other tests makes a file called that. Maybe `test_plotMatrix_testfile.png`?

youngmit replied:
Other option is to do the whole thing in a TemporaryDirectoryChanger, then we dont need to worry about removing anything.

ntouran commented on armi/utils/tests/test_utils.py:
```diff
@@ -250,6 +251,20 @@ def test_convertToSlice(self):
+        xtick = ([0, 1],["1", "2"])
+        ytick = ([0, 1],["1", "2"])
+        utils.plotMatrix(matrix, fname, xticks=xtick, yticks=ytick)
+        os.remove("F.png")
```
consider `f"{fname}.png"` here

ntouran commented on armi/reactor/tests/test_reactors.py:
```diff
@@ -537,6 +545,21 @@ def test_removeAssembly(self):
+        assems = {
+            i: self.r.core.childrenByLocator[loc]
+            for i, loc in enumerate(aLoc)
+            if loc in self.r.core.childrenByLocator.keys()
```
.keys() should be optional there.

ntouran commented on armi/reactor/tests/test_reactors.py:
```diff
@@ -334,6 +334,51 @@ def test_findAllAxialMeshPoints(self):
+            108.7944413504,
+            123.3003668638,
+        ]
+        self.assertListEqual(expectedPoints, aziPoints)
```
We sometimes use things like `np.testing.assert_almost_equal` for list comparisons of floats, which may twiddle sometimes. 

There's also some ongoing concern about how to handle the maintenance burden of tests like this that change when the test fixtures change. We have lots like this so you're doing as we always do, but it's just something we worry about. We'll probably eventually need to decouple more tests from the "one test reactor" loaded in this case, even though we thought it was convenient at first. Just a fyi.

ntouran commented on armi/reactor/tests/test_blocks.py:
```diff
@@ -1339,6 +1339,47 @@ def test_getMfp(self):
+        flux = self.Block.getMgFlux(average=True)
+        referenceFlux = [
+            3.68362183e07,
+            5.21203125e08,
```
can these numbers be derived algorithmically from the dummy data loaded up so we don't have to manually rebuild this if that changes? Or at that point are we just re-implementing the function being tested?

mgjarrett replied:
Probably? I'll look into it. I didn't dig too deep here, but it looks like it's just multiplying the flux and volume provided in the test.

mgjarrett replied:
This test was actually not covering the if block that I wanted it to, so I'm going to delete it. We would need a block.p.lastMgFlux defined in order to test the part of getMgFlux() that I was targeting.

ntouran commented on armi/reactor/tests/test_blocks.py:
```diff
@@ -1607,6 +1607,9 @@ def testSymmetryFactor(self):
         self.assertAlmostEqual(m0 / 3.0, self.HexBlock.getMass())
 
+        symmetryLine = self.HexBlock.isOnWhichSymmetryLine()
+        self.assertEqual(4, symmetryLine)
```
use `grids.BOUNDARY_CENTER` here rather than `4`.

ntouran commented on armi/reactor/tests/test_blocks.py:
```diff
@@ -1708,6 +1708,10 @@ def test_getDuctPitch(self):
 
+    def test_getPinCenterFlatToFlat(self):
+        f2f = self.HexBlock.getPinCenterFlatToFlat()
+        self.assertAlmostEqual(1.3336791218280355, f2f)
```
any option to build this from the input data? Again we've just been bitten a lot when the test fixtures change and we have like N-hundred unit tests to go update it can be a pain.

mgjarrett replied:
To be honest, I wasn't exactly sure what this is calculating, so I just copied the output it produced (fingers crossed that it's already implemented correctly). I'm not sure what "flat-to-flat" means in reference to pins. Is that just the pin pitch? I'll take a look at the numbers and see where this is coming from.

mgjarrett replied:
I changed this to calculate from the input data stored on self.HexBlock. It looks a lot like the function it's testing, but still better for the future if the test fixture changes.

ntouran commented on armi/reactor/blocks.py:
```diff
@@ -2216,7 +2216,7 @@ def thetaInner(self):
         """Return a largest theta of all the components."""
         outerTheta = self.getDimensions("outer_theta")
-        largestOuter = min(outerTheta) if outerTheta else None
+        largestOuter = max(outerTheta) if outerTheta else None
```
Nice catch!!

ntouran commented on armi/reactor/zones.py:
```diff
@@ -599,7 +599,7 @@ def _buildAssemTypeZones(core, cs, typeSpec=None):
             zone = zones[zoneName]
         except KeyError:
-            zone = Zone(a.name)
+            zone = Zone(zoneName)
```
looks right to me. another great catch. thanks!

ntouran commented:
Thanks for all these! Almost a full % increase in coverage! That's the best we've had in a long time. And great bug finds too. 

I went through and made a few small comments and observations. I think @youngmit will want to go through some of these as well. 

One thing about the commit messages is that they should all be in the imperative mood, so "Add a unit test for getAllXsSuffixes." instead of "Adding a unit test for getAllXsSuffixes." This is just a style/consistency thing apparently [traceable back to git man pages. ](https://git.kernel.org/pub/scm/git/git.git/tree/Documentation/SubmittingPatches?id=HEAD#n133). Not a huge deal. 

We'll also have to ask if we want to squash these or not. I kind of like them atomic as you have them TBH but will discuss with @youngmit. Anyway great job.

mgjarrett commented:
I think committing in bits and pieces is good for reviewing, but once they're in the repo it seems unlikely we would ever want to go back to an intermediate state somewhere in the middle of this, so squashing might be the way to go. I'll add a single commit addressing all of your comments.

youngmit reviewed (changes requested):
This all mostly looks great! Just a few things here and there.

youngmit commented:
> 
> 
> I think committing in bits and pieces is good for reviewing, but once they're in the repo it seems unlikely we would ever want to go back to an intermediate state somewhere in the middle of this, so squashing might be the way to go. I'll add a single commit addressing all of your comments.

Wholeheartedly agree. We can squash on the way in

mgjarrett commented on armi/reactor/converters/geometryConverters.py:
```diff
@@ -159,7 +159,8 @@ def convert(self, r=None):
                 else:
                     # non-fuel. Control?
-                    self.setNumberOfBlocks(a, blockType=a[fuelI].getType())
+                    blockType = a[fuelI].getType()
+                    self.setNumberOfBlocks(a, blockType=Flags.fromString(blockType))
```
I made both changes: using `b.p.flags`, and changing the kwarg name to `blockFlags`.

We might want to take a look at what `setNumberOfBlocks` is doing with it. We're passing the flags from the block in the assembly at the index `fuelI`, which is the index of the first fuel block in a reference fuel assembly. `setNumberOfBlocks` then determines the "active" height that needs to be re-meshed:

`fuelHeight = assem.getTotalHeight(blockFlags)`

For a non-fuel assembly, we probably want `blockFlags `to be something like `Flags.CONTROL` or `Flags.SHIELD`. But is it safe to assume `a[fuelI].p.flags` would give us `Flags.CONTROL` and not `Flags.REFLECTOR` or something else "inert"? We're effectively assuming all assemblies are identical up to the beginning of the fuel, right?

youngmit commented on armi/utils/tests/test_utils.py:
```diff
@@ -161,6 +192,84 @@ def test_list2str(self):
+        utils.plotMatrix(matrix, fname, show=True, title="plot")
+        utils.plotMatrix(matrix, fname, minV=0, maxV=5, figsize=[3, 4])
+        utils.plotMatrix(matrix, fname, xticks=xtick, yticks=ytick)
+        dc.__exit__(None, None, None)
```
is this where that os.remove() used to be? the DirectoryChangers are actually [context managers](https://docs.python.org/3/reference/datamodel.html#context-managers), meaning that you can enter them with a `with ...:` statement. Like:
```
with directoryChangers.TemporaryDirectoryChanger(): # __enter__() called implicitly here
    # do stuff
# __exit__() implicitly called here
```

mgjarrett replied:
got it. my initial attempt at this type of usage didn't work how I wanted it, I think because I was using the wrong method. I copied that usage from a setUp/tearDown in a unit test, where enter/exit are called explicitly because they're in different functions. I'll fix this.

youngmit reviewed (changes requested):
One last thing: Use the directory changer as a context manager block

ntouran commented:
horray that got us rounding up to 80% and therefore out of the red!  THANKS!!
![image](https://user-images.githubusercontent.com/53948397/109245032-a1e5f700-7794-11eb-9e9a-a976dbcf8069.png)
